### ISSUE TITLE
enforceHTTPS causes unwanted redirect.

### ISSUE THREAD

**Issue by Crispy1975** [deis/deis#4504]
We have a number of APIs that service an AngularJS application, one of these APIs is legacy and will be replaced in time but for now we proxy requests from the newer API to the legacy one (HapiJS to nginx/PHP/FastCGI). This works perfectly on the Deis platform in both SSL and non-SSL.

However, when we turn on `enforceHTTPS` to make sure that all requests are in HTTPS our legacy API endpoint is exposed to the AngularJS application and we get a CORS error. The CORS error is expected as the legacy API should never have requests made to it directly from the application.

What I believe is happening is that with `enforceHTTPS` set to true on the router the request from the new API to the legacy API is being redirected which in turn exposes the legacy API endpoint. As the redirect only happens with `enforceHTTPS` set to true I think the block of code in the nginx template below is doing the redirect...

https://github.com/deis/deis/blob/v1.9.0/router/image/templates/nginx.conf#L303

We are running v1.9.0 of the Deis platform on AWS using a Cephless install. SSL termination is done on the Deis router as later on we require the use of sockets which AWS ELBs do not support.

All of out API endpoints only ever make HTTPS calls so there should never be a cause for the Deis router to action a redirect.

**Comment by benwilber** [comment]
Hi @Crispy1975 I think there is probably an issue here, but I'm having trouble groking the architecture of your APIs, specifically:

> when we turn on enforceHTTPS to make sure that all requests are in HTTPS our legacy API endpoint is exposed to the AngularJS application and we get a CORS error.

Are you saying that it's *proxying* to the legacy API? Which doesn't set CORS headers, thus the errors.  Are you able to provide some example routes to both APIs that would demonstrate the issue?

**Comment by Crispy1975** [comment]
Essentially our AngularJS SPA only ever make requests to our newest API, it never makes direct calls to our legacy API. However the legacy API endpoint gets exposed to the SPA when enforceHTTPS is on due to what I believe is the Deis router doing aa 301 redirect.

Only a few old routes are "proxied" from our Node.js API to the older PHP one, but I suspect that this proxying with enforcHTTPS on is confusing the router. Perhaps headers sent are not as expected?

In essence as the application proxy request is made in pure HTTPS I would not expect the router to do a redirect at all.

**Comment by Crispy1975** [comment]
I should add to be clear everything works well with `enforceHTTPS` set to `false`.

**Comment by benwilber** [comment]
Just so I'm clear on the flow, is this an accurate depiction of your path:

```
ELB (tcp/443) -> Deis router (https/443) -> nodejs API (http/80) -> PHP API (http/80)
```

**Comment by Crispy1975** [comment]
@benwilber Pretty much, but the final step where the Node.js API calls the PHP API is done with the FQDN as Deis does not yet have internal service discovery. Our API calls are always made to the FQDN over HTTPS, so the route would always be secure and our platform can follow 12 factor methodologies as closely as possible.

As an example the application makes it's call to `https://api.example.com`, that API sees the requested route legacy and in turn makes a request to `https://apiv1.example.com`. This works 100% with `enforceHTTPS` set to `false`.

Based on the route in the flow would look more like this...

    ELB (tcp/443) -> Deis router (https/443) -> nodejs API (http/80) -> ELB (tcp/443) -> Deis router (https/443) -> PHP API (http/80)

So we never make a non-SSL request ever.

**Comment by benwilber** [comment]
@Crispy1975 do curl requests for this work as expected?

`ELB (tcp/443) -> Deis router (https/443) -> PHP API (http/80)`

Since the nodejs API is making requests to the legacy API via the external route, I'm unclear where the redirect is happening.

**Comment by Crispy1975** [comment]
Yes. Direct curl and browser requests (with and author token) work. With enforceHTTPS on I believe the redirect happens between the Node.js request to PHP. This is the only reason the PHP endpoint would be returned to the browser.

The PHP API should just return data to the Node.js API (which made the request) without ever being known to the front end.

I will see if I can get some request and response headers.

**Comment by Crispy1975** [comment]
Ok, here is an example request. What is going on here is our application is making a request to our primary API, as this particular route is still in our older legacy API (PHP) this is proxied like so:

	Application <--> Primary API (old route) <--> PHP API

Here are the requests headers

	Remote Address: xxx.xxx.xxx.xxx:443
	Request URL: https://api.example.com/msg/f344a433-5d83-4810-9bca-6fddf31ce069
	Request Method: GET
	Status Code: 301 Moved Permanently

So, as you see we get a 301 back (redirected). We only ever get this if we have the `enforceHTTPS` set to `true`. If it is set to `false` - or is not set at all - then all requests work perfectly in full `https` with no issues or redirects. Here are the response headers.

	access-control-allow-credentials: true
	access-control-allow-headers: Authorization, Content-Type, If-None-Match
	access-control-allow-methods: GET, HEAD, POST, PUT, PATCH, DELETE, OPTIONS
	access-control-allow-origin: https://app.example.com
	access-control-expose-headers: WWW-Authenticate, Server-Authorization
	access-control-max-age: 86400
	cache-control: no-cache
	content-encoding: gzip
	content-type: text/html
	date: Sun, 25 Oct 2015 08:55:02 GMT
	location: https://apiv1.example.com/msg/f344a433-5d83-4810-9bca-6fddf31ce069
	server: nginx/1.9.4
	status: 301 Moved Permanently
	vary: origin,accept-encoding
	version: HTTP/1.1

For completeness here are the rest of the request headers from Chrome.

	:host: api.example.com
	:method: GET
	:path: /msg/f344a433-5d83-4810-9bca-6fddf31ce069
	:scheme: https
	:version: HTTP/1.1
	accept: application/json, text/plain, */*
	accept-encoding: gzip, deflate, sdch
	accept-language: en-GB,en-US;q=0.8,en;q=0.6
	cookie: xxxxxxx
	origin: https://app.example.com
	referer: https://app.example.com/
	user-agent: Mozilla/5.0 (X11; Linux i686) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/46.0.2490.71 Safari/537.36

To confirm this is 100% using SSL termination at the Deis router(s) and the platform works perfectly well with `enforceHTTPS` set to `false` in either `http` or `https`. We would of course prefer to run pure `https` and not have application hacks to force `https`.

**Comment by krancour** [comment]
The confd template for the router's `nginx.conf` is hard to follow at times.  To easily observe the actual difference between your two scenarios, I would suggest grabbing a copy of `/opt/nginx/nginx.conf` from one of the router containers when `enforceHTTPS ` is disabled and things are working, then do the same with it enabled and have a look at the diff.

**Comment by Crispy1975** [comment]
@krancour thanks, will give that a go and report back here...

**Comment by bacongobbler** [comment]
this should be resolved with #3072.

**Comment by krancour** [comment]
@Crispy1975 this issue has finally been replicated!  I've opened up #4997 to document the root cause.  I think it's ideal to move further conversation to that issue since it's focused on the root cause and how to correct it.  As such, I'm going to close out this one.

**Comment by Crispy1975** [comment]
Cool, thanks @krancour
